### PR TITLE
build: add scripts for privileged docker test execution

### PR DESCRIPTION
**Pull request by klarose**
*Description*: We want to run some integration tests which require various privileges. I have added some scripts which wrap a bazel test target in another script which sets up such a remote environment. The general approach is similar to `bazel-test-gdb`, except no intermediate script is generated.

The script works fairly well. See the documentation portion of the PR for some examples. There are a few gotchas with it, however, that we should probably resolve over time:
 1. It won't work with tests that depend on anything outside the test itself -- e.g. input files.
 2. If compiling locally, there may be some challenges invoking other binaries in the target container, since we end up overwriting some runtime libraries that the container's gdb may need.
 3. There is no strategy just yet to run a privileged test in gdb... I suspect that `sudo bazel-test-gdb` will have to be sufficient.

Example: `tools/bazel-test-docker.sh  //test/integration:integration_test  --jobs=4 -c dbg`

*Risk Level*: Low. This is not used by anything, and it will only ever be used explicitly, so it should be tested substantially then.

*Testing*: Tested locally with various integration tests, as well as in circle with my integration test which requires privileges: https://circleci.com/gh/klarose/envoy-circle/11. This commit was based off my hacked up fork of envoy, which only ran the one test: https://github.com/klarose/envoy-circle

*Docs Changes*: I added a description of the script to bazel/README.md.

*Release Notes*: Only affects tooling, so I don't think one is required.

Fixes Issue #5246

### CODE REVIEW

File: bazel/README.md [hunk 1]
```diff
@@ -200,6 +200,50 @@ tools/bazel-test-gdb //test/common/http:async_client_impl_test -c dbg
...
 binaries will not include debugging symbols and GDB will not be very useful.
 
+# Running Bazel tests requiring privileges
+
+Some tests may require privileges (e.g. CAP_NET_ADMIN) in order to execute. One option is to run
+them with elevated privileges, e.g. `sudo test`. However, that may not always be possible,
+particularly if the test needs to run in a CI pipeline. `tools/bazel-test-docker.sh` may be used in
+such situations to run the tests in a privileged docker container.
+
+The script works by wrapping the test execution in the current repositories circle ci build
```
Comment (zuercher) [inline bazel/README.md]: s/repositories/repository's/

File: tools/bazel-test-docker.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  echo "\"$1\" does not match this pattern"
+  exit 1
+fi
+
+SCRIPT_DIR="$(realpath "$(dirname "$0")")"
+[[ -z "${BAZEL}" ]] && BAZEL=bazel
+[[ -z "${DOCKER}" ]] && DOCKER=docker
+
+if [[ -z "${RUN_REMOTE}" ]]; then
+  LOCAL_MOUNT=${LOCAL_MOUNT:-"yes"}
```
Comment (zuercher) [inline tools/bazel-test-docker.sh]: Should probably quote the assignment: `LOCAL_MOUNT="${LOCAL_MOUNT:-yes}"` here and below.

File: tools/bazel-test-docker.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+# Pass through the docker environment
+DOCKER_ENV=$(mktemp -t docker_env.XXXXXX)
+function cleanup() {
+  rm -f "DOCKER_ENV"
+}
+
+trap cleanup EXIT
+cat > "${DOCKER_ENV}" <<EOF
+  #!/bin/bash
+  export DOCKER_CERT_PATH=$DOCKER_CERT_PATH
```
Comment (zuercher) [inline tools/bazel-test-docker.sh]: Should probably quote/brace the variable references here (e.g. `export DOCKER_CERT_PATH="${DOCKER_CERT_PATH}"`).

File: tools/bazel-test-docker.sh [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  RUN_REMOTE=no
+else
+  LOCAL_MOUNT=${LOCAL_MOUNT:-"no"}
+  RUN_REMOTE=yes
+fi
+
+# Pass through the docker environment
+DOCKER_ENV=$(mktemp -t docker_env.XXXXXX)
+function cleanup() {
+  rm -f "DOCKER_ENV"
```
Comment (zuercher) [inline tools/bazel-test-docker.sh]: `"${DOCKER_ENV}"` to delete the actual file.

File: tools/docker_wrapper.sh [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  LIB_MOUNTS=""
+  if [ "${LOCAL_MOUNT}" == "yes" ]
+  then
+    for path in $LIB_PATHS; do
+      LIB_MOUNTS="${LIB_MOUNTS} -v ${path}:${path}:ro"
+      done
+  fi
+
+# Note: we don't quote LIB_MOUNTS on purpose; we want it to expand.
+docker run --rm --privileged  -v "${TEST_PATH}:/test" $LIB_MOUNTS -i -v "${ENVFILE}:/env" \
```
Comment (zuercher) [inline tools/docker_wrapper.sh]: `${LIB_MOUNTS}` for consistency.

File: tools/docker_wrapper.sh [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  # In this case, we need to create the container, then make new layers on top of it, since we
+  # can't mount everything into it.
+  docker create -t --privileged  --name "${CONTAINER_NAME}" "${IMAGE}" \
+    bash -c "${CMDLINE}"
+  docker cp "$TEST_PATH" "${CONTAINER_NAME}:/test"
+  docker cp "$ENVFILE" "${CONTAINER_NAME}:/env"
+
+  # If some local libraries are necessary, copy them over.
+  if [ "${LOCAL_MOUNT}" == "yes" ]; then
+    for path in $LIB_PATHS; do
```
Comment (zuercher) [inline tools/docker_wrapper.sh]: `${LIB_PATHS}` for consistency.

File: tools/bazel-test-docker.sh (hunk 4 above)
Comment (klarose) [reply tools/bazel-test-docker.sh]: Whoops

File: tools/docker_wrapper.sh [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  LIB_MOUNTS=""
+  if [ "${LOCAL_MOUNT}" == "yes" ]
+  then
+    for path in $LIB_PATHS; do
+      LIB_MOUNTS="${LIB_MOUNTS} -v ${path}:${path}:ro"
+      done
+  fi
+
+# Note: we don't quote LIB_MOUNTS on purpose; we want it to expand.
+docker run --rm --privileged  -v "${TEST_PATH}:/test" ${LIB_MOUNTS} -i -v "${ENVFILE}:/env" \
```
[the remark this answers is not in the document]
Comment (klarose) [reply tools/docker_wrapper.sh]: Whoops. Took me a bit to spot for some reason. Good eye!

**Pull request merged**